src/main/java/org/jenkinsci/plugins/jvctb/config/CredentialsHelper.java

```diff
@@ -1,96 +1,110 @@
 
 public class CredentialsHelper {
-  public static ListBoxModel doFillUsernamePasswordCredentialsIdItems() {
-    final List<StandardUsernamePasswordCredentials> credentials = getAllCredentials();
-    final AbstractIdCredentialsListBoxModel<
-            StandardUsernameListBoxModel, StandardUsernameCredentials>
-        listBoxModel = new StandardUsernameListBoxModel().includeEmptyValue();
-    for (final StandardUsernamePasswordCredentials credential : credentials) {
-      listBoxModel.with(credential);
+  public static ListBoxModel doFillCredentialsIdItems(
+      Item item, String credentialsId, String bitbucketServerUrl) {
+    StandardListBoxModel result = new StandardListBoxModel();
+    if (item == null) {
+      if (!Jenkins.getInstance().hasPermission(Jenkins.ADMINISTER)) {
+        return result.includeCurrentValue(credentialsId);
+      }
+    } else {
+      if (!item.hasPermission(Item.EXTENDED_READ)
+          && !item.hasPermission(CredentialsProvider.USE_ITEM)) {
+        return result.includeCurrentValue(credentialsId);
+      }
     }
-    return listBoxModel;
+    return result //
+        .includeEmptyValue() //
+        .includeMatchingAs(
+            item instanceof Queue.Task ? Tasks.getAuthenticationOf((Queue.Task) item) : ACL.SYSTEM,
+            item,
+            StandardCredentials.class,
+            URIRequirementBuilder.fromUri(bitbucketServerUrl).build(),
+            CredentialsMatchers.anyOf(
+                CredentialsMatchers.instanceOf(StandardUsernamePasswordCredentials.class),
+                CredentialsMatchers.instanceOf(StringCredentials.class)))
+        .includeCurrentValue(credentialsId);
   }
 
-  public static ListBoxModel doFillPersonalAccessTokenIdItems() {
-    final List<StringCredentials> credentials = getAllCredentials(StringCredentials.class);
-    final ListBoxModel listBoxModel =
-        new StandardListBoxModel() //
-            .includeEmptyValue() //
-            .withAll(credentials);
-    return listBoxModel;
-  }
-
-  public static Optional<StandardUsernamePasswordCredentials> findCredentials(
-      final String usernamePasswordCredentialsId) {
-    if (isNullOrEmpty(usernamePasswordCredentialsId)) {
-      return absent();
+  public static FormValidation doCheckFillCredentialsId(
+      Item item, String credentialsId, String bitbucketServerUrl) {
+    if (item == null) {
+      if (!Jenkins.getInstance().hasPermission(Jenkins.ADMINISTER)) {
+        return FormValidation.ok();
+      }
+    } else {
+      if (!item.hasPermission(Item.EXTENDED_READ)
+          && !item.hasPermission(CredentialsProvider.USE_ITEM)) {
+        return FormValidation.ok();
+      }
     }
-
-    return fromNullable(
-        firstOrNull(getAllCredentials(), allOf(withId(usernamePasswordCredentialsId))));
+    if (isNullOrEmpty(credentialsId)) {
+      return FormValidation.ok();
+    }
+    // TODO: We could do a actual check credentials works on the service depending on the API
```

Comment:
Can we do this today with the current violation api?

Reply:
A bit strange wording in that `TODO` comment. I'm not really sure what the question is.

Reply:
In the violation API? Do we support validating credentials? Like `if (login.successful)` do stuff

Reply:
The `violation-comments-to-bitbucket-server-lib` cannot check if given credentials are valid or not. It can just fire some violations with a given credential, and fail hard if the credentials are wrong =)

Reply:
Well then consider the todo removed 😄